# Assignment logistic regression

Logistic regression is a popular classification technique due to its simplicity, interpretability and performance. 

The goal of this assignment is to get insight into the following topics:
- Being able to train a logistic regression model.
- Performing hyperparameter tuning.
- Testing and evaluation of the trained model. Being able to assess the model by means of different classification metrics: accuracy, recall, precision, f1-score.
- Knowing how to deal with underfitting and overfitting.

In [1]:
%matplotlib inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn import preprocessing
from sklearn.preprocessing import PolynomialFeatures
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.linear_model import LogisticRegression
import seaborn as sns # fancy plots

## 1. Wine classification

The dataset *wine_data.csv* consists of chemical parameters of several different wines. Your task is to train a logistic regression model being able to classify these wines by cultivar (wine farmer). In other words can you predict who produced a specific wine?

The column cultivar contains the target values (what you want to predict). For example 1 stands for cultivar 1.

### Reading the dataset and analysis

In [3]:
# Reading the dataset
dataset = pd.read_csv('wine_data.csv')
dataset.head(100)


,inputs Alcohol,MalicAcid,Ash,AlcalinityOfAsh,Magnesium,TotalPhenols,flavanoids,NonflavanoidsPhenols,Proanthocyanins,ColorIntensity,Hue,OD280/OD315,Proline,Cultivar
0,12.08,1.33,2.30,23.6,70,2.20,1.59,0.42,1.38,1.74,1.07,3.21,625,1
1,12.08,1.13,2.51,24.0,78,2.00,1.58,0.40,1.40,2.20,1.31,2.72,630,1
2,12.37,1.17,1.92,19.6,78,2.11,2.00,0.27,1.04,4.68,1.12,3.48,510,1
3,13.11,1.01,1.70,15.0,78,2.98,3.18,0.26,2.28,5.30,1.12,3.18,502,1
4,12.04,4.30,2.38,22.0,80,2.10,1.75,0.42,1.35,2.60,0.79,2.57,580,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,14.34,1.68,2.70,25.0,98,2.80,1.31,0.53,2.70,13.00,0.57,1.96,660,2
96,12.67,0.98,2.24,18.0,99,2.20,1.94,0.30,1.46,2.62,1.23,3.16,450,1
97,12.64,1.36,2.02,16.8,100,2.02,1.41,0.53,0.62,5.75,0.98,1.59,450,1
98,13.20,1.78,2.14,11.2,100,2.65,2.76,0.26,1.28,4.38,1.05,3.40,1050,0


Determine the number of classes. In other words, how many different cultivars are in the dataset? 
Is the dataset balanced? Balanced means that the distribution of the targets is more or less uniformly distributed.
For this dataset balanced means that you have approximately the same number wines for each of the cultivars.
You can use the Seaborn countplot for this. https://seaborn.pydata.org/generated/seaborn.countplot.html.
The problem with imbalanced datasets is that the model you train can get a preference for the majority class.

In [2]:
# Number of classes (countplot)


In [4]:
# Checking for (im)balancedness


Check for missing, extreme or unrealistic values. Remove them when present in the dataset.
You can use the dataset.describe() function to determine the key statistical properties of the dataset.
https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.describe.html). 


In [3]:
# Statistical analysis


In [4]:
# pairplot


### Preprocessing and compiling a training set and test set

Split into **features and targets**.

In [5]:
# Splitting into features and targets


Create a training set and test set. Make sure that 70 wines end up in the test set.
Use the train_test_split function. More information:http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html

In [6]:
# Creating training set and test set


Scale the dataset by means of a Standardscaler or MinMax scaler.
http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html

In [7]:
# Scaling with standardscaler



### Training and testing

Train a logistic regression classifier - (http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html). Initially choose C=1.

Test the trained model on the test set. Compute the confusion matrix, accuracy and generate the classification report. What are your findings?

Try to improve the accuracy of the model by hyperparameter tuning. Options for hyperparameter tuning are the following:
- Find a good C-value
- Choose a different solver
- Apply L1 or L2 regularization (penalty parameter)
- Polynomial expansion of the features: adding higher order features.
- In case you have an imbalanced dataset you can use the **class_weight=balanced** parameter. Explain the effect of it in terms of accuracy , recall, precision and f1-score.





In [8]:
# Training and hyperparameter tuning


In [9]:
# Testing



### Predictions with the trained model

Predict the cultivar for the following wine:

inputs Alcohol: 13.52 - 
MalicAcid: 2.05 - 
Ash: 2.20 - 
AlcalinityOfAsh: 17.3 - 
Magnesium: 120 - 
TotalPhenols: 2.60 - 
flavanoids: 3.52 - 
NonflavanoidsPhenols: 0.30 - 
Proanthocyanins: 2.28 - 
ColorIntensity: 7.80 - 
Hue: 0.77 - 
OD280/OD315: 2.90 - 
Proline: 862


What is the confidence of the model? Use the predict_proba() function to compute this value.


In [10]:
# Predictions


Look at the output of 'lregmodel.coef_' and 'lregmodel.intercept_', where lregmodel is the name of the logistic regression model. Your model might have a different name. So replace lregmodel with the name of the model you used. 

You will see that you have seperate coefficients for each class.
What does a large positve coefficient mean? And what does a negative coefficient mean?


In [11]:
# Interpretation of the coefficients


## 2. Cancer detection

The file 'cancer.csv' consists of data about several tumors. The diagnosis variable is what you try to predict. M stands for a malignant tumor, B for a benign tumor.
Train a logistic regression classifier to make a an as accurately as possible diagnosis




In [17]:
# reading the dataset
dataset = pd.read_csv('cancer.csv')
print('Dimension of the dataset:',dataset.shape)
dataset.head()

Dimension of the dataset: (568, 32)


,id,diagnosis,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,...,radius_worst,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst
0,842517,M,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
1,84300903,M,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
2,84348301,M,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
3,84358402,M,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678
4,843786,M,12.45,15.70,82.57,477.1,0.12780,0.17000,0.1578,0.08089,...,15.47,23.75,103.40,741.6,0.1791,0.5249,0.5355,0.1741,0.3985,0.12440


In [13]:
# Check the balancedness of the data


# Remove the last column and the id column


# Replace B and M by 0 and 1 in diagnosis column


# Split into features and targets


# Make a training set and a test set. Make sure that 1/3 of the dataset is in the test set.


# Scale the features



Train a logistic regression 

Test the models on the test set. Which one performs best?
Does it make sense to expand the dataset with extra features? 

http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.PolynomialFeatures.html

In [43]:
# Split into features and targets



In [44]:
# Split into training set and test set



In [45]:
# Scaling with MinMax scaler


In [11]:
# Train and test you logistic regression classifier


In [12]:
# Optimize the logistic regression classifier with hyperparameter tuning



You are being asked to reduce the number of false negatives on the test set to zero. False negatives in this example means: prediciton is benign while in fact the tumor is malignant.
To achieve this, you can vary the threshold on the predict_prob() to make the classifier more sensitive for the class 'malignant'. Try to keep the accuracy as high as possible.
Discuss the results.

## 50K a year

Train a classifier (logistic regressoin or SVM) to predict whether or not a person earns more than 50000 dollar a year. Do hypertuning to each the highest possbile accuracy. Also evaluate by means of recall, precision, f1-score.



In [16]:
dataset = pd.read_csv('50K_a_year.csv')
dataset.head()

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,class
0,25,Private,226802,11th,7,Never-married,Machine-op-inspct,Own-child,Black,Male,0,0,40,United-States,<=50K
1,38,Private,89814,HS-grad,9,Married-civ-spouse,Farming-fishing,Husband,White,Male,0,0,50,United-States,<=50K
2,28,Local-gov,336951,Assoc-acdm,12,Married-civ-spouse,Protective-serv,Husband,White,Male,0,0,40,United-States,>50K
3,44,Private,160323,Some-college,10,Married-civ-spouse,Machine-op-inspct,Husband,Black,Male,7688,0,40,United-States,>50K
4,18,?,103497,Some-college,10,Never-married,?,Own-child,White,Female,0,0,30,United-States,<=50K


In [15]:
# 50K a year 



## 4. Activity tracking

The dataset 'activity.csv' consists of smartphone measurements of 30 different people. The measuments come from the built-in gyroscopes and accelerometers of these smartphones. With each of the measurements in the dataset corresponds a human activity (standing, walking, walking upstairs, ...). In total, there are 6 different activities /classes.


Train a logistic regression classifier to predict the human activity as accurately as possible. In order to get the accuracy as high as possible you can apply hyperparameter tuning and/or feature engineering (higher order features, feature selection, etc)

Write down some conclusions about the performance of your model. Does you model have difficulties with certain classes? Is you model suffering from imbalancedness? How did you improve the performance of the model?
Did you apply regularization?


In [38]:
dataset = pd.read_csv('activity.csv')
dataset.head(100)

,tBodyAcc-mean()-X,tBodyAcc-mean()-Y,tBodyAcc-mean()-Z,tBodyAcc-std()-X,tBodyAcc-std()-Y,tBodyAcc-std()-Z,tBodyAcc-mad()-X,tBodyAcc-mad()-Y,tBodyAcc-mad()-Z,tBodyAcc-max()-X,...,fBodyBodyGyroJerkMag-kurtosis(),"angle(tBodyAccMean,gravity)","angle(tBodyAccJerkMean),gravityMean)","angle(tBodyGyroMean,gravityMean)","angle(tBodyGyroJerkMean,gravityMean)","angle(X,gravityMean)","angle(Y,gravityMean)","angle(Z,gravityMean)",subject,Activity
0,0.288585,-0.020294,-0.132905,-0.995279,-0.983111,-0.913526,-0.995112,-0.983185,-0.923527,-0.934724,...,-0.710304,-0.112754,0.030400,-0.464761,-0.018446,-0.841247,0.179941,-0.058627,1,STANDING
1,0.278419,-0.016411,-0.123520,-0.998245,-0.975300,-0.960322,-0.998807,-0.974914,-0.957686,-0.943068,...,-0.861499,0.053477,-0.007435,-0.732626,0.703511,-0.844788,0.180289,-0.054317,1,STANDING
2,0.279653,-0.019467,-0.113462,-0.995380,-0.967187,-0.978944,-0.996520,-0.963668,-0.977469,-0.938692,...,-0.760104,-0.118559,0.177899,0.100699,0.808529,-0.848933,0.180637,-0.049118,1,STANDING
3,0.279174,-0.026201,-0.123283,-0.996091,-0.983403,-0.990675,-0.997099,-0.982750,-0.989302,-0.938692,...,-0.482845,-0.036788,-0.012892,0.640011,-0.485366,-0.848649,0.181935,-0.047663,1,STANDING
4,0.276629,-0.016570,-0.115362,-0.998139,-0.980817,-0.990482,-0.998321,-0.979672,-0.990441,-0.942469,...,-0.699205,0.123320,0.122542,0.693578,-0.615971,-0.847865,0.185151,-0.043892,1,STANDING
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,0.327317,-0.022256,-0.149144,-0.248645,0.133229,-0.179547,-0.306979,0.113994,-0.159712,-0.114953,...,-0.422166,-0.438911,-0.710035,0.121018,-0.023408,-0.769477,0.256167,0.038575,1,WALKING
96,0.349059,-0.022004,-0.150052,-0.303125,0.152607,-0.174298,-0.356618,0.106512,-0.141309,-0.178067,...,-0.493845,-0.538717,0.833201,0.457682,0.211527,-0.776736,0.251406,0.036180,1,WALKING
97,0.264682,-0.008592,-0.102974,-0.363238,0.036867,-0.244159,-0.425625,-0.055041,-0.243230,-0.278328,...,-0.569276,0.460655,0.765162,0.424757,0.167463,-0.786963,0.244471,0.034677,1,WALKING
98,0.284317,-0.027206,-0.212303,-0.183444,0.141076,-0.225689,-0.249080,0.110614,-0.221009,0.032522,...,-0.551872,-0.053158,0.512810,-0.811518,0.581624,-0.764185,0.260565,0.031442,1,WALKING


In [ ]:
# Activity tracking
